src/caffe/net.cpp

```diff
@@ -650,6 +650,61 @@ void Net<Dtype>::ShareTrainedLayersWith(const Net* other) {
 }
 
 template <typename Dtype>
+void Net<Dtype>::ShareBlobsWith(const Net* other) {
+  int num_source_blobs = other->blobs().size();
+  for (int i = 0; i < num_source_blobs; ++i) {
+    const string& source_blob_name = other->blob_names()[i];
+    if (
+      std::find(
+        other->input_blob_indices().begin(),
+        other->input_blob_indices().end(), i) !=
+      other->input_blob_indices().end()) {
+      // input blob cannot be shared
+      LOG(INFO) << "Ignoring source blob " << source_blob_name;
+      continue;
+    }
+    int target_blob_id = 0;
+    while (target_blob_id != blob_names_.size() &&
+      blob_names_[target_blob_id] != source_blob_name) {
+      ++target_blob_id;
+    }
+    if (target_blob_id == blob_names_.size() ||
+      std::find(
+        net_input_blob_indices_.begin(),
+        net_input_blob_indices_.end(), target_blob_id) !=
+      net_input_blob_indices_.end()) {
+      LOG(INFO) << "Ignoring source blob " << source_blob_name;
+      continue;
+    }
+    LOG(INFO) << "Sharing source blob " << source_blob_name;
+    // We do not care blob shapes because every call of `Forward` method
+    // reshapes output blobs into right shapes, except input blobs.
+    blobs_[target_blob_id] = other->blobs()[i];
```

Comment:
Yeah, first I tried to do like that. However, I noticed that `ShareData` actually do sharing `SyncedMemory` instance (shared_ptr), and if you reshape the blob after sharing it and the allocated capacity is not large enough as you need for reshaping, it will create a new synced memory instance. Then, the coupling will be gone. This will happen in the special case that an input layer produces different shape blobs for each data. That is why I decide to share blobs directly. But, yes, your point is right. Using `UpdateBlobPointers` makes it a little complicated. If you have any better idea to implement this, please help me. Thanks!